machineconfiguration/v1/types_machineosconfig.go

```diff
@@ -0,0 +1,207 @@
+type MachineOSConfigStatus struct {
+	// observedGeneration represents the generation of the MachineOSConfig object observed by the Machine Config Operator's build controller.
+	// +kubebuilder:validation:XValidation:rule="self >= oldSelf", message="observedGeneration must not move backwards"
+	// +kubebuilder:validation:Minimum=0
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+	// currentImagePullSpec is the fully qualified image pull spec used by the MCO to pull down the new OSImage. This includes the sha256 image digest.
+	// This is generated when the Machine Config Operator's build controller successfully completes the build.
+	// The format of the image pull spec is: host[:port][/namespace]/name@sha256:<digest>,
+	// where the digest must be 64 characters long, and consist only of lowercase hexadecimal characters, a-f and 0-9.
+	// The length of the whole spec must be between 1 to 447 characters.
+	// +optional
+	CurrentImagePullSpec ImageDigestFormat `json:"currentImagePullSpec,omitempty"`
+	// machineOSBuild is a reference to the MachineOSBuild object for this MachineOSConfig, which contains the status for the image build.
+	// +optional
+	MachineOSBuild *ObjectReference `json:"machineOSBuild,omitempty"`
+	// conditions are state related conditions for the object.
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	// +optional
+	// TODO(jerzhang): add godoc after conditions are finalized.
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+}
+
+type MachineOSImageBuilder struct {
+	// imageBuilderType specifies the backend to be used to build the image.
+	// +kubebuilder:validation:Enum:=Job
+	// Valid options are: Job
+	// +required
+	ImageBuilderType MachineOSImageBuilderType `json:"imageBuilderType"`
+}
+
+// MachineOSContainerfile contains all custom content the user wants built into the image
+type MachineOSContainerfile struct {
+	// containerfileArch describes the architecture this containerfile is to be built for.
+	// This arch is optional. If the user does not specify an architecture, it is assumed
+	// that the content can be applied to all architectures, or in a single arch cluster: the only architecture.
+	// +kubebuilder:validation:Enum:=ARM64;AMD64;PPC64LE;S390X;NoArch
+	// +kubebuilder:default:=NoArch
+	// +optional
+	ContainerfileArch ContainerfileArch `json:"containerfileArch,omitempty"`
+	// content is an embedded Containerfile/Dockerfile that defines the contents to be built into your image.
+	// See https://github.com/containers/common/blob/main/docs/Containerfile.5.md for the spec reference.
+	// for example, this would add the tree package to your hosts:
+	//   FROM configs AS final
+	//   RUN rpm-ostree install tree && \
+	//     ostree container commit
+	// +required
+	// +kubebuilder:validation:MaxLength=4096
```

Comment:
A small sentence explaining the max length is required at the end of the godoc please